src/boot/efi/boot.c

```diff
@@ -1162,6 +1271,35 @@ static void config_add_entry(Config *config, BootEntry *entry) {
         config->entries[config->n_entries++] = entry;
 }
 
+static void config_add_group(Config *config, BootGroup *group) {
+        assert(config);
+        assert(group);
+
+        assert(config->n_groups < IDX_MAX);
+
+        config->groups = xrealloc(
+                        config->groups,
+                        sizeof(void *) * config->n_groups,
+                        sizeof(void *) * (config->n_groups + 1));
+
+        config->groups[config->n_groups++] = group;
+}
+
+static void group_add_entry(BootGroup *group, BootEntry *entry) {
+        assert(group);
+        assert(entry);
+
+        assert(group->n_entries < IDX_MAX);
+
+        if ((group->n_entries & 15) == 0) {
```

Comment:
i am not too much of a fan of such a logic. makes things fragile, in case we start out with a an array initialized some unexpected way...

also feels like premature optimization to me?

also, you could as well store the actually allocated size in another field?